### PR TITLE
run,performance: check build stage status in `truncated exponential backoff` manner

### PR DESCRIPTION
**Pull request by flouthoc**
Buildah constantly polls running build container for it status every `0.1` second, this polling nature wastes CPU cycles unnecessarily so instead of checking status every `0.1` second use exponentially increasing counter.

Exponential counter has a max limit of `2 seconds` and delay will reset to certain milliseconds and will grow again to `2 seconds` this process will be repeated till status is returned from the build container.

[NO NEW TESTS NEEDED]
[NO TESTS NEEDED]
This is not a new feature but improvement to buildah so it stops consuming and producing unnecessary overhead for jobs with longer duration.

Closes: https://github.com/containers/buildah/issues/4481

#### What type of PR is this?

> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test 
> /kind feature
> /kind flake
> /kind other

#### Does this PR introduce a user-facing change?

```release-note
run,performance: check build stage status in `exponential backoff` manner
```

### CODE REVIEW

File: run_common.go [hunk 1]
```diff
@@ -642,7 +644,15 @@ func runUsingRuntime(options RunOptions, configureNetwork bool, moreCreateArgs [
 		}
 	}()
 	signal.Notify(interrupted, syscall.SIGHUP, syscall.SIGINT, syscall.SIGTERM)
+	iter := 0
 	for {
+		iter = iter + 1
+		exponential := int(math.Min((math.Pow(2, float64(iter)) + rand.Float64()*500), 2000))
```
Comment (rhatdan) [inline run_common.go:650]: I don't see this as an exponential back off. more of a random timer?

I wrote a test program using your logic and I saw:
```
Inter= 1 304
Inter= 2 474
Inter= 3 340
Inter= 4 234
Inter= 5 244
Inter= 6 407
Inter= 7 160
Inter= 8 334
Inter= 9 560
Inter= 10 1174
Inter= 11 2000
```
I would have expected
```
Inter= 1 100
Inter= 2 200
Inter= 3 400
Inter= 4 800
Inter= 5 1600
Inter= 6 2000
```

Comment (rhatdan) [reply run_common.go:650]: In my example, I don't see the interval growing it goes from .407 seconds to .160 seconds on the next try.

Comment (flouthoc) [reply run_common.go]: In random it grows but in non-uniform manner and there are dips in between but maybe I am overthinking the random part hence  I have removed the random part and now pattern should look like.

```
-----iter 100
-----iter 200
-----iter 400
-----iter 800
-----iter 1600
-----iter 2000
-----iter 100
-----iter 200
-----iter 400
-----iter 800
-----iter 1600
-----iter 2000
```

Comment (rhatdan) [reply run_common.go]: I think that works well.

**Comment by rhatdan** [comment]
LGTM
@vrothberg @giuseppe PTAL

File: run_common.go [hunk 2]
```diff
@@ -642,7 +643,15 @@ func runUsingRuntime(options RunOptions, configureNetwork bool, moreCreateArgs [
 		}
 	}()
 	signal.Notify(interrupted, syscall.SIGHUP, syscall.SIGINT, syscall.SIGTERM)
+	iter := 0
 	for {
+		iter = iter + 1
+		exponential := int(math.Min((math.Pow(2, float64(iter)) * 50), 2000))
+		// Reached the max delay time reset counter to 0
+		// and start exponential delay again.
+		if exponential == 2000 {
```
Comment (rhatdan) [inline run_common.go:652]: I am not sure of this code here, should we just stay at 2 seconds once it gets here rather then speeding up again?

**Comment by giuseppe** [comment]
this will make the build process slower, since it takes longer to realize the container terminated.

`crun state` is really fast, I don't see how it is a problem to run it 10 times per second

**Comment by flouthoc** [comment]
> this will make the build process slower, since it takes longer to realize the container terminated.
> 
> `crun state` is really fast, I don't see how it is a problem to run it 10 times per second

I think in original issue the concern was made for longer steps where build container is definitely gonna take time and even then buildah somehow keeps polling in background , hogging the resource via (exec calls). But still waiting for reply from @paravz on original issue https://github.com/containers/buildah/issues/4481 though maybe he is facing different issue than this one.

**Comment by giuseppe** [comment]
> I think in original issue the concern was made for longer steps where build container is definitely gonna take time and even then buildah somehow keeps polling in background , hogging the resource via (exec calls). But still waiting for reply from @paravz on original issue #4481 though maybe he is facing different issue than this one.

`crun state` is still cheap though.  I don't think it is a good idea to make the entire process slower just to reduce the number of fork/exec calls

**Comment by rhatdan** [comment]
Is there other ways we could check?  IE Watch for a exit code to appear?

**Comment by flouthoc** [comment]
> Is there other ways we could check? IE Watch for a exit code to appear?

@rhatdan No I think the actual issue was polling , but once actual problem is explained in detail and if it gets decided that its needed then maybe we can make exponential backoff configurable via a flag. So just waiting for discussion here: https://github.com/containers/buildah/issues/4481

**Pull request closed without merging**